pkg/cloud/rgraph/exec/executor_parallel.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2024 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/rgraph/algo"
+	"k8s.io/klog/v2"
+)
+
+var (
+	ErrPendingActions = errors.New("Executor did not process all actions")
+)
+
+func defaultParallelExecutorConfig() *ExecutorConfig {
+	return &ExecutorConfig{
+		DryRun:        false,
+		ErrorStrategy: ContinueOnError,
+		Timeout:       5 * time.Minute,
+	}
+}
+
+// NewParallelExecutor returns a new Executor that runs tasks multi`-threaded.
+func NewParallelExecutor(pending []Action, opts ...Option) (*parallelExecutor, error) {
+	ret := &parallelExecutor{
+		config: defaultParallelExecutorConfig(),
+		result: &Result{Pending: pending},
+		pq:     algo.NewParallelQueue[Action](),
+	}
+	for _, opt := range opts {
+		opt(ret.config)
+	}
+
+	if err := ret.config.validate(); err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+type parallelExecutor struct {
+	config *ExecutorConfig
+	c      cloud.Cloud
+
+	// lock guards results
+	lock   sync.Mutex
+	result *Result
+
+	pq   *algo.ParallelQueue[Action]
+	done chan *TraceEntry
+}
+
+// parallelExecutor implements Executor.
+var _ Executor = (*parallelExecutor)(nil)
+
+func (ex *parallelExecutor) Run(ctx context.Context, c cloud.Cloud) (*Result, error) {
+	ex.queueRunnableActions()
+	subctx, cancel := context.WithTimeout(ctx, ex.config.Timeout)
+
+	runAction := func(ctx context.Context, a Action) error {
```

Comment:
It would be good to just take this out and make it a method.

func (ex *parallelExecutor) runAction(ctx, a) {...}

Reply:
I created this function here because I wanted to capture the cloud variable.
Actually I think it would be better to pass cloud in constructor "NewParallelExecutor" function and change Run to take only context. 
But this would require changes in Executor:
```
type Executor interface {
	// Run the actions. Returns non-nil if there was an error in execution of
	// one or more Actions.
	Run(context.Context) (*Result, error)
}
```

Reply:
done